File: DatabindingWithNewsApi/app/src/main/java/com/enpassio/databindingwithnewsapi/viewmodel/MainViewModel.java
package com.enpassio.databindingwithnewsapi.viewmodel;

import android.app.Application;
import android.arch.lifecycle.AndroidViewModel;
import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.content.Intent;
import android.databinding.ObservableField;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;

import com.enpassio.databindingwithnewsapi.model.Article;
import com.enpassio.databindingwithnewsapi.repository.NewsRepository;
import com.enpassio.databindingwithnewsapi.utils.NetworkUtils;
import com.enpassio.databindingwithnewsapi.utils.UIState;

import java.util.List;

import static android.support.constraint.Constraints.TAG;

public class MainViewModel extends AndroidViewModel {

    private LiveData<List<Article>> mArticleList;
    private Article mChosenArticle;
    private final NewsRepository mRepo;
    private MutableLiveData<Boolean> showSnack = new MutableLiveData<>();
    public final ObservableField<UIState> uiState = new ObservableField<>(UIState.LOADING);

    public MainViewModel(@NonNull Application application) {
        super(application);
        //Passing the application context to the repository (not activity context!)
        mRepo = NewsRepository.getInstance();
        showSnack.setValue(false);
    }

    public LiveData<List<Article>> getArticleList() {
        if (mArticleList == null) {
            mArticleList = mRepo.getArticles();
            checkConnectionAndStartLoading();
        }
        return mArticleList;
    }

    public Article getChosenArticle() {
        return mChosenArticle;
    }

    public void setChosenArticle(Article chosenArticle) {
        mChosenArticle = chosenArticle;
    }

    public LiveData<Boolean> shouldShowSnack() {
        return showSnack;
    }

    /*If there is internet connection, start fetching data from the internet,
   otherwise show a snack for warning user*/
    public void checkConnectionAndStartLoading() {
        if (NetworkUtils.thereIsConnection(getApplication())) {
            /*If there is connection, start fetching and change uiState
            to LOADING. This will show a loading indicator*/
            uiState.set(UIState.LOADING);
            mRepo.startFetching();
            showSnack.setValue(false);
        } else {
            //If there is no connection, show a snack message to warn user
            showSnack.setValue(true);
            /*Unless there is some previously fetched data to show,
            set uiState to NETWORK_ERROR This will show an error message*/
            if(mArticleList.getValue() == null || mArticleList.getValue().isEmpty()){
                uiState.set(UIState.NETWORK_ERROR);
            }
        }
    }
    public void openWebSite(View view) {
        if (mChosenArticle == null) {
            return;
        }
        String articleUrl = mChosenArticle.getArticleUrl();
        Uri webUri = null;
        if (!TextUtils.isEmpty(articleUrl)) {
            //Parse string to uri
            try {
                webUri = Uri.parse(articleUrl);
            } catch (Exception e) {
                Log.e(TAG, e.toString());
            }
            //Send an implicit intent to open the article in the browser
            Intent webIntent = new Intent(Intent.ACTION_VIEW);
            webIntent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
            webIntent.setData(webUri);
            if (webIntent.resolveActivity(view.getContext().getPackageManager()) != null) {
                view.getContext().startActivity(webIntent);
            }
        }
    }
}

File: DatabindingWithNewsApiKotlin/app/src/main/java/com/enpassio/databindingwithnewsapikotlin/viewmodel/MainViewModel.kt
package com.enpassio.databindingwithnewsapikotlin.viewmodel

import android.app.Application
import android.arch.lifecycle.AndroidViewModel
import android.arch.lifecycle.LiveData
import android.arch.lifecycle.MutableLiveData
import com.enpassio.databindingwithnewsapikotlin.data.Article
import com.enpassio.databindingwithnewsapikotlin.data.NewsRepository
import com.enpassio.databindingwithnewsapikotlin.utils.UIState
import com.enpassio.databindingwithnewsapikotlin.utils.thereIsConnection

class MainViewModel(application: Application) :
    AndroidViewModel(application) {

    /*UI state keeps track of the data loading state: LOADING, NETWORK_ERROR or SUCCESS
    This information is kept in an observable field inside view model, so that
    changes are automatically reflected in UI. */
    val uiState = ObservableField<UIState>(UIState.LOADING)

    /*If article list is not null, return it, otherwise claim the
   data from the repository, assign it to articleList and start fetching*/
    var articleList: LiveData<List<Article>?>? = null
        get() = field ?: NewsRepository.articles.also {
            field = it
            checkConnectionAndStartLoading()
        }

    //Chosen article will be set later when an item is selected from the list
    var chosenArticle: Article? = null

    //If there is no network and data can't be fetched, show a snackbar to the user
    private val _showSnack = MutableLiveData<Boolean>()
    val showSnack : LiveData<Boolean>
        get() = _showSnack

    init {
        _showSnack.value = false
    }

    /*If there is internet connection, start fetching data from the internet,
    otherwise show a snack for warning user*/
    fun checkConnectionAndStartLoading() {
        if (thereIsConnection(getApplication())) {
            /*If there is connection, start fetching and change uiState
            to LOADING. This will show a loading indicator*/
            uiState.set(UIState.LOADING)
            NewsRepository.startFetching()
            _showSnack.value = false
        } else {
            //If there is no connection, show a snack message to warn user
            _showSnack.value = true
            /*Unless there is some previously fetched data to show,
            set uiState to NETWORK_ERROR This will show an error message*/
            if(articleList?.value.isNullOrEmpty()){
                uiState.set(UIState.NETWORK_ERROR)
            }
        }
    }

    companion object {
        private const val TAG = "MainViewModel"
    }

}